File: LeetCode-C++/LeetCode-C++/二叉树/_98_验证二叉搜索树.cpp
//
//  _98_验证二叉搜索树.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/11/25.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/validate-binary-search-tree/

#include <stdio.h>
#include <stack>
#include <climits>
#include <limits.h>
#include "TreeNode.hpp"

using namespace std;

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        TreeNode* p = root;
        stack<TreeNode *> st;
        long inorder = LONG_MIN;
        
        while (p || !st.empty()) {
            while (p) {
                st.push(p);
                p = p->left;
            }
            p = st.top();
            st.pop();
            if (p->val <= inorder) return false;
            inorder = p->val;
            p = p->right;
        }

        return true;
    }
};

/**
 示例 1:

 输入:
     2
    / \
   1   3
 输出: true
 示例 2:

 输入:
     5
    / \
   1   4
      / \
     3   6
 输出: false
 解释: 输入为: [5,1,4,null,null,3,6]。
      根节点的值为 5 ，但是其右子节点值为 4 。
 */
/**
int main(int argc, const char * argv[]) {
//    TreeNode* root = generateBinaryTree({2, 1, 3});
    TreeNode* root = generateBinaryTree({5,1,4,INT_MAX,INT_MAX,3,6});
    Solution solution = Solution();
    printf("%d\n", solution.isValidBST(root));
    return 0;
}
*/

File: LeetCode-Swift/LeetCode-Swift/二叉树/_98_验证二叉搜索树.swift
//
//  _98_验证二叉搜索树.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/11/25.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/validate-binary-search-tree/

import Foundation

class _98_验证二叉搜索树 {
    
    func isValidBST(_ root: TreeNode?) -> Bool {
        var p = root;
        var nodes = [TreeNode]()
        
        var inorder = Int.min
        while p != nil || nodes.count > 0 {
            while p != nil {
                nodes.append(p!)
                p = p!.left
            }
            p = nodes.popLast()
            guard p!.val > inorder else { return false }
            
            inorder = p!.val
            p = p?.right
        }
        
        return true
    }
    
    var last: Int?
    func isValidBST1(_ root: TreeNode?) -> Bool {
        if root == nil {
            return true;
        }
        if !isValidBST(root?.left) {
            return false
        }
        if (last != nil && (last! > root!.val)) {
            return false
        }
        last = root?.val
        if !isValidBST(root?.right) {
            return false
        }
        return true
    }
}

/**
 输入:
     2
    / \
   1   3
 输出: true
 */
func test_98() {
    let root = TreeNode(2)
    root.left = TreeNode(1)
    root.right = TreeNode(3)
    let solution = _98_验证二叉搜索树()
    print(solution.isValidBST(root))
}
